src/Language/Fixpoint/Solver/PLE.hs

```diff
@@ -875,13 +1018,80 @@ evalApp γ ctx e0 es et
         else do
           return (Nothing, noExpand)
 
+evalApp _ ctx e0 es _
+  | EVar f <- dropECst e0
+  , Just rewrite <- M.lookup f (icSMTRewrites ctx)
+  = do
+    -- No need to check if the eta beta flag is on since icSMTRewrites is emoty
+    modify $ \st ->
+      st { evNewEqualities = S.insert (eApps e0 es, eApps rewrite es) (evNewEqualities st) }
+    return (Just $ eApps rewrite es, expand)
+
+evalApp _ ctx e0 es _
+  | deANFed <- deANF ctx e0
+  , dropECst deANFed /= dropECst e0
+  = do
+      isEtaBetaOn <- gets etabetaFlag
+      if isEtaBetaOn then
+        return (Just $ eApps deANFed es, expand)
+      else
+        return (Nothing, noExpand)
+
+-- **ETA**: PLE before the patch to perform eta expansion was only able to unfold
+-- fully applied functions, by eta equivalence we mean the property that for all
+-- functions f the terms f and (\x -> f x) have the same semantics, we leverage 
+-- this fact to fully apply functions with extra function arguments so to trigger
+-- the standard PLE espansion, see the following example:
+-- We have a function f:
+-- define f (x : int) : int = {(x)}
+-- And we need to prove some constraint of this shape
+-- { f = something }
+-- The original PLE cannot do anything on f as it's not fully applied, instead
+-- if we perform eta expansion on f we obtain the following constraint
+-- { \y:Int -> f y = something }
+-- And now PLE can perform the unfolding of f as it's now fully saturated
+-- { \y:Int -> y = something }
+-- Clearly we need the higerorder flag active as we are generating lambda in
+-- the constraints.
+evalApp _γ _ctx e0 es _et
+  | ECst (EVar _f) sortAnnotation@FFunc{} <- e0
+  = do
+    isEtaBetaOn <- gets etabetaFlag
+    let expectedArgs = init $ flatten sortAnnotation
+    let nProvidedArgs = length es
+    let nArgsMissing = length expectedArgs - nProvidedArgs
+    if isEtaBetaOn && nArgsMissing > 0 then do
+      let etaArgsType = drop nProvidedArgs $ flatten sortAnnotation
+      -- Fresh names for the eta expansion
+      etaNames <- makeFreshEtaNames nArgsMissing
+
+      let etaVars = zipWith (\name ty -> ECst (EVar name) ty) etaNames etaArgsType
+      let fullBody = eApps e0 (es ++ etaVars)
+      let etaExpandedTerm = mkLams fullBody (zip etaNames etaArgsType)
+
+      -- Note: we should always add the equality as etaNames is always non empty because the
+      -- only way for etaNames to be empty is if the function is fully applied, but that case
+      -- is already handled by the previous case of evalApp
+      modify $ \st -> st 
+        { evNewEqualities = S.insert (eApps e0 es, etaExpandedTerm) (evNewEqualities st) }
+      return (Just etaExpandedTerm, expand)
+    else do
+      pure (Nothing, noExpand)
+  where
+    flatten (FFunc t ts) = t : flatten ts
+    flatten t = [t]
```

Comment:
How about
```Haskell
    unpackFFuncs :: Sort -> ([Sort], Sort)
    unpackFFuncs (FFunc t ts) =
      let (args, ret) = unpackFFuncs ts
       in (t : args, ret)
    unpackFFuncs t = ([], t)
```
It would avoid the init and drop calls above.